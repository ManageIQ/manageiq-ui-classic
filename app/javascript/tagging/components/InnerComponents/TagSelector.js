import React from 'react';
import PropTypes from 'prop-types';
import { Dropdown } from 'carbon-components-react';
import TaggingPropTypes from '../TaggingPropTypes';

class TagSelector extends React.Component {
  // eslint-disable-next-line react/sort-comp
  handleChange = (val) => {
    const { onTagCategoryChange, tagCategories } = this.props;
    const selectedOption = val.selectedItem;
    if (selectedOption) {
      onTagCategoryChange({
        id: selectedOption.value,
        description: tagCategories.find((category) => category.id === selectedOption.value).description,
      });
    }
  };

  labelWithIcon = (description, infoText) => (
    <div>
      <span
        style={{
          display: 'inline-block',
          width: 'calc(100% - 18px)',
        }}
      >
        {description}
      </span>
      <span
        className="pull-right pficon pficon-info tag-icon"
        title={infoText}
        aria-hidden="true"
      />
      <span className="sr-only">{infoText}</span>
    </div>
  );

  // eslint-disable-next-line react/destructuring-assignment
  tagCategories = this.props.tagCategories.map((category) => ({
    keyWord: category.description.toLowerCase(),
    value: category.id,
    label: category.singleValue
      // eslint-disable-next-line react/destructuring-assignment
      ? this.labelWithIcon(category.description, this.props.infoText)
      : category.description,
  }));

  getOptions = (values) => {
    const options = [];
    values.forEach((item) => (
      options.push(
        { key: item.value, value: item.value, label: item.label }
      )
    ));
    return options;
  }

  render() {
    const { selectedOption, isDisabled } = this.props;
    let value = selectedOption.id ? { label: selectedOption.description, value: selectedOption.id } : null;
    if (value === null) {
      value = { value: -1 };
    }
    return (
      <Dropdown
        className="tag-select"
        id="dropdown-tag-select"
        label={`${__('Select tag category')}`}
        disabled={isDisabled}
        onChange={(val) => this.handleChange(val)}
        items={this.getOptions(this.tagCategories)}
      />
    );
  }
}
TagSelector.propTypes = {
  tagCategories: PropTypes.arrayOf(TaggingPropTypes.category).isRequired,
  selectedOption: TaggingPropTypes.value,
  onTagCategoryChange: PropTypes.func.isRequired,
  infoText: PropTypes.string,
  isDisabled: PropTypes.bool,
};

TagSelector.defaultProps = {
  infoText: __('Only a single value can be assigned from these categories'),
  selectedOption: {},
  isDisabled: false,
};

export default TagSelector;
