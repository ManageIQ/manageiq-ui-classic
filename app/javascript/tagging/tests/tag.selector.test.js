import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import TagSelector from '../components/InnerComponents/TagSelector';

describe('Tagging modifier', () => {
  const tagCategories = [
    { description: 'Name', id: 1 },
    { description: 'Number', id: 2 },
  ];
  const selectedTagValue = { description: 'Homer', id: 1 };
  let onTagCategoryChange;
  beforeEach(() => {
    onTagCategoryChange = jest.fn();
  });

  it('match snapshot', () => {
    const tree = shallow(<TagSelector
      tagCategories={tagCategories}
      onTagCategoryChange={onTagCategoryChange}
      selectedOption={selectedTagValue}
    />);
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('should call methods', () => {
    const wrapper = shallow(
      <TagSelector
        tagCategories={tagCategories}
        onTagCategoryChange={onTagCategoryChange}
        selectedOption={selectedTagValue}
      />
    );
    wrapper.instance().handleChange({ selectedItem: { label: 'Name', value: 1 } });
    expect(onTagCategoryChange.mock.calls).toHaveLength(1);
  });
});
