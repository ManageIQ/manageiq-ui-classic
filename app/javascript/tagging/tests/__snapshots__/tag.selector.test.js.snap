// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tagging modifier match snapshot 1`] = `
<Dropdown
  className="tag-select"
  direction="bottom"
  disabled={false}
  helperText=""
  id="dropdown-tag-select"
  itemToElement={null}
  itemToString={[Function]}
  items={
    Array [
      Object {
        "key": 1,
        "label": "Name",
        "value": 1,
      },
      Object {
        "key": 2,
        "label": "Number",
        "value": 2,
      },
    ]
  }
  label="Select tag category"
  light={false}
  onChange={[Function]}
  titleText=""
  type="default"
/>
`;
